pvlib/iotools/era5.py

```diff
@@ -0,0 +1,294 @@
+    else:
+        area = [latitude+0.005, longitude-0.005,
+                latitude-0.005, longitude+0.005]
+
+    params = {
+        'product_type': product_type,
+        'format': 'netcdf',
+        'variable': variables,
+        'date': start.strftime('%Y-%m-%d') + '/' + end.strftime('%Y-%m-%d'),
+        'time': ERA5_HOURS,
+        'grid': grid,
+        'area': area}
+
+    # Retrieve path to the file
+    file_location = cds_client.retrieve(dataset, params)
+
+    # Load file into memory
+    with requests.get(file_location.location) as res:
+
+        # Save the file locally if local_path has been specified
+        if save_path is not None:
+            with open(save_path, 'wb') as f:
+                f.write(res.content)
+
+        return read_era5(res.content, map_variables=map_variables,
+                         output_format=output_format)
+
+
+def read_era5(filename, output_format=None, map_variables=True):
+    """Read one or more ERA5 netcdf files.
+
+    Parameters
+    ----------
+    filename: str or path-like or list
+        Filename of a netcdf file containing ERA5 data or a list of filenames.
+    output_format: {'dataframe', 'dataset'}, optional
+        Type of data object to return. Default is to return a pandas DataFrame
+        if file only contains one location and otherwise return an xarray
+        dataset.
+    map_variables: bool, default: True
+        When true, renames columns to pvlib variable names where applicable.
+        See variable ERA5_VARIABLE_MAP.
+
+    Returns
+    -------
+    data: DataFrame
+        ERA5 time-series data, fields depend on the requested data. The
+        returned object is either a pandas DataFrame or an xarray Dataset, see
+        the output_format parameter.
+    metadata: dict
+        Metadata for the time-series.
+
+    See Also
+    --------
+    pvlib.iotools.get_era5
+
+    References
+    ----------
+    .. [1] `ERA5 hourly data on single levels from 1979 to present
+       <https://cds.climate.copernicus.eu/cdsapp#!/dataset/reanalysis-era5-single-levels?tab=overview>`_
+    .. [2] `ERA5 data documentation
+       <https://confluence.ecmwf.int/display/CKB/ERA5%3A+data+documentation>`_
+    """
+    # open multiple-files - requires dask
+    if isinstance(filename, (list, tuple)):
+        ds = xr.open_mfdataset(filename)
+    else:
+        ds = xr.open_dataset(filename)
+
+    metadata = extract_metadata_from_dataset(ds)
+
+    if map_variables:
+        # Renaming of xarray datasets throws an error if keys are missing
+        ds = ds.rename_vars(
+            {k: v for k, v in ERA5_VARIABLE_MAP.items() if k in list(ds)})
+
+    if (output_format == 'dataframe') or (
+            (output_format is None) & (ds['latitude'].size == 1) &
+            (ds['longitude'].size == 1)):
+        if (ds['latitude'].size == 1) & (ds['longitude'].size == 1):
+            data = ds.to_dataframe().droplevel(['latitude', 'longitude'])
+        return data, metadata
```

Comment:
Not sure this is the logic we want -- `data` isn't necessarily defined here.  What do you think about raising an error if dataframe is requested but the sizes aren't equal to 1?

Reply:
It should only have been the droplevel that was in the if statement and not the to_dataframe(). It looks like this now:

    if (output_format == 'dataframe') or (
            (output_format is None) & (ds['latitude'].size == 1) &
            (ds['longitude'].size == 1)):
        data = ds.to_dataframe()
        if (ds['latitude'].size == 1) & (ds['longitude'].size == 1):
            data.droplevel(['latitude', 'longitude'])
        return data, metadata

You can still get a DataFrame with multiple locations (and multiindex) but then you shouldn't remove the latitude/longitude.